### ISSUE TITLE
Publish to verdaccio thorws EPBULSIHCONFLICT if package already exists at npmjs

### ISSUE THREAD

**Issue by dkebler** [verdaccio/verdaccio#1203]
**Describe the bug**

When publishing to verdaccio a package for the first time (like when u have a new verdaccio install/storage) if that package already exists in npmjs npm will throw an error
```
npm ERR! code EPUBLISHCONFLICT
npm ERR! publish fail Cannot publish over existing version.
```
like here I have my scoped packages at npmjs
https://www.npmjs.com/org/uci
and if I try to publish any of these to a clean verdaccio I get the conflict.

**To Reproduce**

publish a package to npmjs `npm publish --registry https://registry.npmjs.org --access public`
now do a fresh install, or scrub the /storage folder and re add a user.
now `npm publish --registry http://localhost:4873'`

The tarball is copied to storage fine but no entry is made in db.json file and in the log you get the ` EPUBLISHCONFLICT` error

now edit `config.yaml` and remove `proxy: npmjs` from the `packages:` section and the issue goes away. 

**Expected behavior**

Verdaccio shouldn't check/proxy through to npmjs at all when using `npm publish --registry http://localhost:4873'`   That being the whole point right?  If I publish to or ask for a repo and verdaccio is handling it then it should ignore npmjs all together.  See below cause it's maybe not doing that but throwning a stream write error that makes it look so.

**Versions**
I tried various versions including 4.0 beta and This happens for any version going back <3.0 via docker or via global install.

check out #questions at discord for my many posts tracking this down.

**NOTE**  if you do use `npm publish --registry http://localhost:4873 --force`  you can get verdaccio/npm to accept the publish without error.  But that's not a solution as I need it to proxy other packages.

below you see log from vercaddio when first I publish.  the first publish throws a stream error which is probably the real issue which then elicits the confusing EPUBLISHCONFLICT in npm.  

```
 http <-- 409, user: david(127.0.0.1), req: 'PUT /@uci%2fi2c-device', error: this package is already present
 http <-- 409, user: david(127.0.0.1), req: 'PUT /@uci%2fi2c-device', error: this package is already present
 http <-- 200, user: david(127.0.0.1), req: 'GET /@uci%2fi2c-device?write=true', bytes: 0/2956
 http <-- 200, user: david(127.0.0.1), req: 'GET /@uci%2fi2c-device?write=true', bytes: 0/2956
 http <-- 409, user: david(127.0.0.1), req: 'PUT /@uci%2fi2c-device', error: this package is already present
 http <-- 409, user: david(127.0.0.1), req: 'PUT /@uci%2fi2c-device', error: this package is already present
Error [ERR_STREAM_WRITE_AFTER_END]: write after end
    at writeAfterEnd (_stream_writable.js:248:12)
    at WriteStream.Writable.write (_stream_writable.js:296:5)
    at UploadTarball.ondata (_stream_readable.js:693:20)
    at UploadTarball.emit (events.js:188:13)
    at UploadTarball.Readable.read (_stream_readable.js:491:10)
    at flow (_stream_readable.js:957:34)
    at resume_ (_stream_readable.js:938:3)
    at processTicksAndRejections (internal/process/next_tick.js:76:17)
Error: ENOENT: no such file or directory, rename '/home/david/.local/share/verdaccio/storage/@uci/i2c-device/i2c-device-0.1.4.tgz.tmp-0420165471909848' -> '/home/david/.local/share/verdaccio/storage/@uci/i2c-device/i2c-device-0.1.4.tgz'
```

**Comment by dkebler** [comment]
grrrr.  I have one of 11 packages/repos that even with 
`npm publish --registry http://localhost:4873 --force`
will not publish. not even with
`npm unpublish --registry http://localhost:4873 --force && npm publish --registry http://localhost:4873'`

here is the log for that
https://download.kebler.net/2019-01-24T21_20_44_504Z-debug.log


one off behaviour like this is frustrating

**Comment by dkebler** [comment]
more input

for the offending package repo which is 0.1.6 on npmjs
https://www.npmjs.com/package/@uci/i2c-bus

if I bump the version in package.json from 0.1.6 to 0.1.7
and  
`npm publish --registry http://localhost:4873 --force`
it will publish
then if I reset it back to 0.1.6 in package.json and try again
`npm publish --registry http://localhost:4873 --force`
no more error!!

what up with that.  Well at least at this point I have all my library published to verdaccio and maybe can continue but this took me a day+

**Comment by NetOperatorWibby** [comment]
I'm running into this issue as well. I don't like having to needlessly version bump just to get this to work.

**Comment by dekadentno** [comment]
> I'm running into this issue as well. I don't like having to needlessly version bump just to get this to work.
> 
> EDIT: Like OP, I just installed Verdaccio and I ran into this issue with my first publish.
> EDIT 2: `--force` works but I wish I didn't have to do it.
> EDIT 3: Specifying the registry works without the usage of `--force`. So, `npm publish --registry https://registry.npmjs.org` for default and `npm publish --registry https://site.tld` for Verdaccio.

I had situations where I had to do 2 or 3 version bumps to be able to publish it. Not happy about it 😞

**Comment by juanpicado** [comment]
> Verdaccio shouldn't check/proxy through to npmjs at all when using `npm publish --registry http://localhost:4873'` That being the whole point right? If I publish to or ask for a repo and verdaccio is handling it then it should ignore npmjs all together. 

No, if we allow that then the whole proxy approach does not make sense. You can disable hit main registry just creating a specific package section for those packages you want to avoid hit any remote registry. 

```
packages:
  'lodash':
    # scoped packages
    access: $authenticated
    publish: $authenticated
    # proxy: npmjs <-- disable this
```

If you have the `proxy` enabled, it will hit the remote. That is what is for. 
Relates #1106

**Comment by dkebler** [comment]
sadly yarn does not support --force so can't try out my theory

**BUT it dawned on me!  to try adding** `--access restricted`  

**TADA** no more publish conflict!

`npm publish --registry https:// --force --access restricted`

Since at npmjs you can do as you please for private repos, overwriting and back publishing,  adding this 'allows' npm to continue.

Albiet now it's a private repo on my own repository but who cares it's just my dev repository and only I need it.  So works for me since I have a token in .npmrc for installs. 

My guess is that now npm simply does NOT allow --force on any public repos not matter the registry.  Before I suppose it only checked to see if the registry was npmjs.  Now it doesn't care.  This would at least explain why no issue before and this is the fix now.

**Comment by dekadentno** [comment]
> What versions of npm are you all running?? I run the latest always. 6.7 right now.

My npm is 6.7.0

> **BUT it dawned on me! to try adding** `--access restricted`
> 
> **TADA** no more republish conflict! @dekadentno

Good catch there, thank you very much. **I can confirm that this solution is working so far, but I've tested it only on 3-4 packages.**
I'll analyze your solution more thoroughly asap.

But then again, I've never found a logical link between packages that didn't make any problems when publishing and those packages that were giving our famous EPUBLISHCONFLICT issue.

Thanks for your answer, once again @dkebler

**Comment by dekadentno** [comment]
I'll have to bump this closed issue and I apologize if it's against any rules.

Few minutes ago, I was trying to publish [babel-core](https://www.npmjs.com/package/@babel/core) to my private npm repository. After running `npm publish`, I got the infamous EPUBLISHCONFLICT error. Also, after trying with `npm publish --registry http://192.168.xx.xx --force --access restricted`, I got the very same error. Of course, that package doesn't exist on my repository, not anything related to babel.

What I found out is the content of the `.npmignore` file:
```
🍔  matej@Mac-Pro:(master) ~/Desktop/npmshit/babel/packages/babel-core
>cat .npmignore
src
test
*.log
```

After deleting the `.npmignore` file, `npm publish` worked without an issue.

Can anyone help me understand why would anyone put the `src` directory in the `.npmignore`?
